### PR TITLE
Chapter 4 solutions

### PR DESCRIPTION
**Pull request by alejandrogarin**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +485,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure a = Reward a
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Trap f <*> _ = Trap f
+    Reward f <*> a = fmap f a
```
Comment (vrom911) [inline src/Chapter4.hs:492]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -489,6 +503,41 @@ Implement the 'Applicative' instance for our 'List' type.
...
+{-
+-- Examples
+>>> c1 = Cons 1 (Cons 2 Empty)
+>>> f1 = Cons (+1) (Cons (+2) (Cons ((+3)) Empty))
+>>> f1 <*> c1
+Cons 2 (Cons 3 (Cons 3 (Cons 4 (Cons 4 (Cons 5 Empty)))))
+-}
+
+lappend :: List a -> List a -> List a
+lappend Empty Empty = Empty
```
Comment (vrom911) [inline src/Chapter4.hs:523]: This case is actually would be caught by the following two, so can be removed 🙂

File: src/Chapter4.hs [hunk 3]
```diff
@@ -489,6 +503,41 @@ Implement the 'Applicative' instance for our 'List' type.
...
+lappend xs Empty = xs
+lappend Empty xs = xs
+lappend (Cons x Empty) (Cons y Empty) = Cons x (Cons y Empty)
+lappend (Cons x xs) (Cons y Empty) = Cons x (radd xs y)
```
Comment (vrom911) [inline src/Chapter4.hs:527]: These two also can be skipped and only the next one considered 👌🏼

File: src/Chapter4.hs [hunk 4]
```diff
@@ -489,6 +503,41 @@ Implement the 'Applicative' instance for our 'List' type.
...
+lappend (Cons x xs) (Cons y ys) = ladd x (lappend xs (ladd y ys))
+
+ladd :: a -> List a -> List a
+ladd x Empty = Cons x Empty
+ladd x (Cons y Empty) = Cons x (Cons y Empty)
```
Comment (vrom911) [inline src/Chapter4.hs:532]: This case is covered by the next more generic one 🙂

File: src/Chapter4.hs [hunk 5]
```diff
@@ -489,6 +503,41 @@ Implement the 'Applicative' instance for our 'List' type.
...
+ladd x (Cons y ys) = Cons x (Cons y ys)
+
+radd :: List a -> a -> List a
+radd Empty x = Cons x Empty
+radd (Cons y Empty) x = Cons x (Cons y Empty)
+radd (Cons y ys) x  = Cons x (Cons y ys)
```
Comment (vrom911) [inline src/Chapter4.hs:538]: Isn't this function is the same as `ladd` but with the arguments flipped?
You can specify it like this:

```haskell
radd l x = ladd x l
```

or use `flip`:

```haskell
radd = flip ladd
```

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask @chshersh or me any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

<hr>

If you would like to continue your journey to Haskell, check out our [Haskell Knowledge Map](https://kowainik.github.io/images/Haskell_Knowledge_Map.png) for the list of topics to explore 🗺️

File: src/Chapter4.hs (hunk 5 above)
Comment (alejandrogarin) [reply src/Chapter4.hs:538]: 👍

**Pull request merged**